### DISCUSSION TITLE
How to get the `run` and `debug` icons on a custom compiler plugin generated main entry point

### DISCUSSION THREAD

**Discussion by soronpo** [scalameta/metals#6729] [General]
I have compiler plugin that from an annotated class like `@top class Foo` creates a main entry point named `top_Foo`. 
I wonder if there is a way to get metals to recognize the main entry point and place the `run` and `debug` icons above the `@top` annotation (scalac of course recognizes these entry points successfully). Is there something that the compiler plugin needs to do for Metals to be able to recognize this?

**Comment by tgodzik** [comment]
As long as it's found as a main class by the build tool it should work. I guess the problem here is that we do not know where `top_Foo` should be shown, since it's not a class or method that exists in the sourcecode.

**Reply by soronpo** [reply]
Thanks for replying. Can you perhaps take a look and see if there is something that can be done on my end to make this work:
```scala
//> using scala 3.5.0
//> using dep io.github.dfianthdl::dfhdl::0.7.0
//> using plugin io.github.dfianthdl:::dfhdl-plugin:0.7.0
//> using option -deprecation -language:implicitConversions

import dfhdl.* 

@top class Foo extends EDDesign
```

**Reply by tgodzik** [reply]
Looks like we have two problems:
- Bloop doesn't seem to send back _empty_/top_Foo. main class back to Metals (I am not sure currently why)
- _empty_/top_Foo. seems to not have an occurrence in a file, which means we would need a custom logic, or show the code lenses at the top of the file

The first issue is bigger because it blocks running the main class at all. Scala CLI doesn't seem to have an issue here, but that might be due to the fact that it doesn't run via Bloop.

**Reply by soronpo** [reply]
> Looks like we have two problems:
> 
>     * Bloop doesn't seem to send back _empty_/top_Foo. main class back to Metals (I am not sure currently why)
> 
>     * _empty_/top_Foo. seems to not have an occurrence in a file, which means we would need a custom logic, or show the code lenses at the top of the file
> 
> 
> The first issue is bigger because it blocks running the main class at all. Scala CLI doesn't seem to have an issue here, but that might be due to the fact that it doesn't run via Bloop.

Could it be that it's because I'm creating a new symbol without proper positioning? How did you diagnose the issue? I can try to fix it on my end if I can properly compare regular `@main` and my `@top` using the diagnosis.

**Reply by tgodzik** [reply]
We have a custom logic for `@main` due to the same issue actually https://github.com/scalameta/metals/blob/main/metals/src/main/scala/scala/meta/internal/metals/debug/DebugDiscovery.scala#L439

With positions we could possibly make that work, but I am not sure. I usually do `Metals: Show semanticdb detailed` to get the semanticdb doc that we base the code lenses on.

**Reply by tgodzik** [reply]
This would be the one for you file:
```
m.scala
-------

Summary:
Schema => SemanticDB v4
Uri => m.scala
Text => empty
Language => Scala
Symbols => 5 entries
Occurrences => 6 entries
Synthetics => 33 entries

Symbols:
_empty_/Foo# => @top class Foo extends EDDesign { self: Foo => +1 decls }
  top => dfhdl/top#
  EDDesign => dfhdl/core/EDDesign#
  Foo => _empty_/Foo#
_empty_/Foo#`<init>`(). => primary ctor <init>(): Foo
  Foo => _empty_/Foo#
_empty_/top_Foo. => final object top_Foo extends Object with DFApp { self: top_Foo.type => +1 decls }
  Object => java/lang/Object#
  DFApp => dfhdl/app/DFApp#
  top_Foo => _empty_/top_Foo.
local0 => protected method __dfcDFC <: dfhdl/core/Container#__dfc().
  DFC => dfhdl/core/DFC#
local1 => @top final class $anon extends Foo { self: $anon => +1 decls }
  top => dfhdl/top#
  Foo => _empty_/Foo#
  $anon => local1

Occurrences:
[5:7..5:12) => dfhdl/
[7:1..7:4) => dfhdl/top#
[7:4..7:4) <= local1
[7:11..7:14) <= _empty_/Foo#
[7:23..7:23) <= _empty_/Foo#`<init>`().
[7:23..7:31) => dfhdl/hdl$package.EDDesign#

Synthetics:
... a lot of synthetics
```

**Reply by soronpo** [reply]
OK, that helps a lot! I added the `coord` to the symbol. Didn't change anything.

For the following file:
```scala
package demo2
import dfhdl.*

@top class IDTop extends DFDesign

@main def IDTopMain: Unit = {}
```

I get:
```
IDTop.scala
-----------

Summary:
Schema => SemanticDB v4
Uri => IDTop.scala
Text => empty
Language => Scala
Symbols => 8 entries
Occurrences => 10 entries
Synthetics => 33 entries

Symbols:
demo2/IDTop# => @top class IDTop extends DFDesign { self: IDTop => +1 decls }
demo2/IDTop#`<init>`(). => primary ctor <init>(): IDTop
demo2/IDTop$package. => final package object demo2 extends Object { self: demo2.type => +2 decls }
demo2/IDTop$package.IDTopMain(). => @main method IDTopMain=> Unit
demo2/IDTopMain# => final class IDTopMain extends Object { self: IDTopMain => +2 decls }
demo2/top_IDTop. => final object top_IDTop extends Object with DFApp { self: top_IDTop.type => +1 decls }
local0 => protected method __dfcDFC <: dfhdl/core/Container#__dfc().
local1 => @top final class $anon extends IDTop { self: $anon => +1 decls }

Occurrences:
[0:8..0:13) <= demo2/
[1:7..1:12) => dfhdl/
[3:1..3:4) => dfhdl/top#
[3:4..3:4) <= local1
[3:11..3:16) <= demo2/IDTop#
[3:25..3:25) <= demo2/IDTop#`<init>`().
[3:25..3:33) => dfhdl/hdl$package.DFDesign#
[5:1..5:5) => scala/main#
[5:10..5:19) <= demo2/IDTop$package.IDTopMain().
[5:21..5:25) => scala/Unit#

Synthetics:
[3:0..3:4) => *(default(given_LogLevel(given_LogLevel), given_OnError(given_OnError), given_DefaultClkCfg, given_DefaultRstCfg, given_PrintDesignCodeAfter), default(given_CommitFolder, given_NewFolderForTop, given_Backend, given_LogLevel(given_LogLevel), given_PrintDesignCodeAfter, given_PrintGenFiles, given_DropUserOpaques), default(given_Align, given_Color, given_ShowGlobals, given_DesignPrintFilter), default(given_DefaultMode))
[3:0..3:4) => orig().apply[top]
[3:0..3:4) => *(default(given_LogLevel(given_LogLevel), given_OnError(given_OnError), given_DefaultClkCfg, given_DefaultRstCfg, given_PrintDesignCodeAfter), default(given_CommitFolder, given_NewFolderForTop, given_Backend, given_LogLevel(given_LogLevel), given_PrintDesignCodeAfter, given_PrintGenFiles, given_DropUserOpaques), default(given_Align, given_Color, given_ShowGlobals, given_DesignPrintFilter), default(given_DefaultMode))
[3:0..3:4) => *(default(given_LogLevel(given_LogLevel), given_OnError(given_OnError), given_DefaultClkCfg, given_DefaultRstCfg, given_PrintDesignCodeAfter), default(given_CommitFolder, given_NewFolderForTop, given_Backend, given_LogLevel(given_LogLevel), given_PrintDesignCodeAfter, given_PrintGenFiles, given_DropUserOpaques), default(given_Align, given_Color, given_ShowGlobals, given_DesignPrintFilter), default(given_DefaultMode))
[3:0..3:4) => *(default(given_LogLevel(given_LogLevel), given_OnError(given_OnError), given_DefaultClkCfg, given_DefaultRstCfg, given_PrintDesignCodeAfter), default(given_CommitFolder, given_NewFolderForTop, given_Backend, given_LogLevel(given_LogLevel), given_PrintDesignCodeAfter, given_PrintGenFiles, given_DropUserOpaques), default(given_Align, given_Color, given_ShowGlobals, given_DesignPrintFilter), default(given_DefaultMode))
[3:4..3:4) => *(given_LogLevel(given_LogLevel), given_OnError(given_OnError), given_DefaultClkCfg, given_DefaultRstCfg, given_PrintDesignCodeAfter)
[3:4..3:4) => *(given_LogLevel)
[3:4..3:4) => *(given_OnError)
[3:4..3:4) => *(given_CommitFolder, given_NewFolderForTop, given_Backend, given_LogLevel(given_LogLevel), given_PrintDesignCodeAfter, given_PrintGenFiles, given_DropUserOpaques)
[3:4..3:4) => *(given_LogLevel)
[3:4..3:4) => *(given_Align, given_Color, given_ShowGlobals, given_DesignPrintFilter)
[3:4..3:4) => *(given_DefaultMode)
[3:4..3:4) => *(given_LogLevel(given_LogLevel), given_OnError(given_OnError), given_DefaultClkCfg, given_DefaultRstCfg, given_PrintDesignCodeAfter)
[3:4..3:4) => *(given_LogLevel)
[3:4..3:4) => *(given_OnError)
[3:4..3:4) => *(given_CommitFolder, given_NewFolderForTop, given_Backend, given_LogLevel(given_LogLevel), given_PrintDesignCodeAfter, given_PrintGenFiles, given_DropUserOpaques)
[3:4..3:4) => *(given_LogLevel)
[3:4..3:4) => *(given_Align, given_Color, given_ShowGlobals, given_DesignPrintFilter)
[3:4..3:4) => *(given_DefaultMode)
[3:4..3:4) => *(given_LogLevel(given_LogLevel), given_OnError(given_OnError), given_DefaultClkCfg, given_DefaultRstCfg, given_PrintDesignCodeAfter)
[3:4..3:4) => *(given_LogLevel)
[3:4..3:4) => *(given_OnError)
[3:4..3:4) => *(given_CommitFolder, given_NewFolderForTop, given_Backend, given_LogLevel(given_LogLevel), given_PrintDesignCodeAfter, given_PrintGenFiles, given_DropUserOpaques)
[3:4..3:4) => *(given_LogLevel)
[3:4..3:4) => *(given_Align, given_Color, given_ShowGlobals, given_DesignPrintFilter)
[3:4..3:4) => *(given_DefaultMode)
[3:4..3:4) => *(given_LogLevel(given_LogLevel), given_OnError(given_OnError), given_DefaultClkCfg, given_DefaultRstCfg, given_PrintDesignCodeAfter)
[3:4..3:4) => *(given_LogLevel)
[3:4..3:4) => *(given_OnError)
[3:4..3:4) => *(given_CommitFolder, given_NewFolderForTop, given_Backend, given_LogLevel(given_LogLevel), given_PrintDesignCodeAfter, given_PrintGenFiles, given_DropUserOpaques)
[3:4..3:4) => *(given_LogLevel)
[3:4..3:4) => *(given_Align, given_Color, given_ShowGlobals, given_DesignPrintFilter)
[3:4..3:4) => *(given_DefaultMode)
```
According to the code you linked, indeed there is a special case for `scala/main#`. So maybe if I figure out how to inject this on my end it would work like usual?

**Reply by soronpo** [reply]
OK, the problem may lay in how Metals detects objects with `def main`. Since what I have in my library is
```scala
package dfhdl.app
trait DFApp:
  def main(commandArgs: Array[String]): Unit = ???
```
In my plugin I do essentially do `object top_Foo extends app.DFApp`.
So I attempted to this manually and Metals does recognize it.
It's unclear to me why Metals does not detect this when it's done in the plugin:
```scala
package demo2
import dfhdl.*

object Foo extends app.DFApp

@top class IDTop extends DFDesign
```

```
IDTop.scala
-----------

Summary:
Schema => SemanticDB v4
Uri => IDTop.scala
Text => empty
Language => Scala
Symbols => 6 entries
Occurrences => 10 entries
Synthetics => 33 entries

Symbols:
demo2/Foo. => final object Foo extends Object with DFApp { self: Foo.type => +1 decls }
demo2/IDTop# => @top class IDTop extends DFDesign { self: IDTop => +1 decls }
demo2/IDTop#`<init>`(). => primary ctor <init>(): IDTop
demo2/top_IDTop. => final object top_IDTop extends Object with DFApp { self: top_IDTop.type => +1 decls }
local0 => protected method __dfcDFC <: dfhdl/core/Container#__dfc().
local1 => @top final class $anon extends IDTop { self: $anon => +1 decls }

Occurrences:
[0:8..0:13) <= demo2/
[1:7..1:12) => dfhdl/
[3:7..3:10) <= demo2/Foo.
[3:19..3:22) => dfhdl/app/
[3:23..3:28) => dfhdl/app/DFApp#
[5:1..5:4) => dfhdl/top#
[5:4..5:4) <= local1
[5:11..5:16) <= demo2/IDTop#
[5:25..5:25) <= demo2/IDTop#`<init>`().
[5:25..5:33) => dfhdl/hdl$package.DFDesign#

...
```

**Reply by tgodzik** [reply]
I think we might do some checks to make sure that symbol exists in the source. It would probably best id we could get that in the synthetics section and just check that.

**Reply by soronpo** [reply]
> I think we might do some checks to make sure that symbol exists in the source. It would probably best id we could get that in the synthetics section and just check that.

You mean in Metals?

**Reply by tgodzik** [reply]
In semanticdb, which is currently in the compiler itself.

**Reply by soronpo** [reply]
So I should open an issue in the Scala 3 compiler repo?
In any case, since Metals can see the symbol in the SemanticDB, why can't it extract the required information from it (a symbol with a main method declaration)?

**Reply by tgodzik** [reply]
We don't know which symbol has the main method declaration. We can try to figure it out using the presentation compiler, but we don't want to do too many calculations for code lenses, since they are requested very often. The current approach is to only read a single semanticdb file and get all the information from it.

If it's a normal main class then we know based on the build server information, which classes are those. Otherwise, we need some way of knowing that a given annotation corresponds to a new synthetic main class. We could try to check every class but that would be very resource intensive on every change to a file.

I think the biggest issue here is that we don't even get the information about the new synthetic main class from the build server. If we had that information we could try doing some best effort lenses placement or some additional logic.

**Reply by soronpo** [reply]
> I think the biggest issue here is that we don't even get the information about the new synthetic main class from the build server. If we had that information we could try doing some best effort lenses placement or some additional logic.

OK, I'll start to track this by opening an issue in the bloop repo.

**Reply by soronpo** [reply]
Done https://github.com/scalacenter/bloop/issues/2413

**Reply by tgodzik** [reply]
Ok, as mentioned in the issue it seem I can no longer reproduce so it seems I might have done something wrong.

I wonder if we could find the exact location using presentation compiler. So if we have a main class that is not in semanticdb occurrences, we could try to do go to definition and show it there.

**Reply by soronpo** [reply]
> Ok, as mentioned in the issue it seem I can no longer reproduce so it seems I might have done something wrong.

I don't quite understand.
```scala
//> using scala 3.5.0
//> using dep io.github.dfianthdl::dfhdl::0.7.1
//> using plugin io.github.dfianthdl:::dfhdl-plugin:0.7.1

import dfhdl.*
@top class Foo extends EDDesign

object Bar extends app.DFApp
```
For the code above I get:
![image](https://github.com/user-attachments/assets/62eb0259-43f8-4a8d-8f53-a1a2967e2243)

The run/debug icon is only available for `Bar` but not above `Foo`. Can you not reproduce this?

**Reply by tgodzik** [reply]
Some further experiments resulted in:
https://github.com/scalameta/metals/compare/main...tgodzik:metals:experiment-custom?expand=1

What we can do currently is just add any unknown mains (not knowing what produced them) to the top of the file, which will be rather limited effort and impact on performance. (so we can do it soon)

As a possible follow up we could try to add position to `.info` method result, which might be able to give us the exact position if you attach the position to the created symbol. Assuming I understand how this works. We could try to do all the steps separately and if they do not work, it should not be an issue for anything else.

Opinions?

**Reply by soronpo** [reply]
> What we can do currently is just add any unknown mains (not knowing what produced them) to the top of the file, which will be rather limited effort and impact on performance. (so we can do it soon)

Slight ambiguity if there is more than one `@top` in the same file. The user can be instructed to avoid that, so this could be a solution, at least short term. 

> if you attach the position to the created symbol

From what I was able to ascertain, from v0.7.1 the symbol has proper positioning.

> We could try to do all the steps separately and if they do not work, it should not be an issue for anything else.

Sounds like a good plan. I really appreciate it you looking into this.

**Reply by tgodzik** [reply]
> I don't quite understand.

Sorry, I just meant the Bloop issue. Code lenses are still an issue.

I quickly got something like this working:

![example-scala-3_5_0-1725034699840](https://github.com/user-attachments/assets/47e083f9-e6f1-4d6e-a7ce-04cab397a398)

Also, I should refresh the page before writing a response :sweat_smile:

**Reply by soronpo** [reply]
> > I don't quite understand.
> 
> Sorry, I just meant the Bloop issue. Code lenses are still an issue.
> 
> I quickly got something like this working:
> 
> ![example-scala-3_5_0-1725034699840](https://private-user-images.githubusercontent.com/3807253/363257115-47e083f9-e6f1-4d6e-a7ce-04cab397a398.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUwMzYxMjUsIm5iZiI6MTcyNTAzNTgyNSwicGF0aCI6Ii8zODA3MjUzLzM2MzI1NzExNS00N2UwODNmOS1lNmYxLTRkNmUtYTdjZS0wNGNhYjM5N2EzOTguZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MzBUMTYzNzA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjEyZDc2YmY0Y2QwYWM3MDdjMDVlYmVhMDg1ZDk2ZTg3NzIxMjIxNWM3YjdiZDhiNjg0M2VmMWFjNDY0Y2M5YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.FiGRgj20XfFj11euX0dmt-DmJ9DlCGltswvbXn6sHf8) [ ![example-scala-3_5_0-1725034699840](https://private-user-images.githubusercontent.com/3807253/363257115-47e083f9-e6f1-4d6e-a7ce-04cab397a398.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUwMzYxMjUsIm5iZiI6MTcyNTAzNTgyNSwicGF0aCI6Ii8zODA3MjUzLzM2MzI1NzExNS00N2UwODNmOS1lNmYxLTRkNmUtYTdjZS0wNGNhYjM5N2EzOTguZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MzBUMTYzNzA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjEyZDc2YmY0Y2QwYWM3MDdjMDVlYmVhMDg1ZDk2ZTg3NzIxMjIxNWM3YjdiZDhiNjg0M2VmMWFjNDY0Y2M5YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.FiGRgj20XfFj11euX0dmt-DmJ9DlCGltswvbXn6sHf8) ](https://private-user-images.githubusercontent.com/3807253/363257115-47e083f9-e6f1-4d6e-a7ce-04cab397a398.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUwMzYxMjUsIm5iZiI6MTcyNTAzNTgyNSwicGF0aCI6Ii8zODA3MjUzLzM2MzI1NzExNS00N2UwODNmOS1lNmYxLTRkNmUtYTdjZS0wNGNhYjM5N2EzOTguZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MzBUMTYzNzA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjEyZDc2YmY0Y2QwYWM3MDdjMDVlYmVhMDg1ZDk2ZTg3NzIxMjIxNWM3YjdiZDhiNjg0M2VmMWFjNDY0Y2M5YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.FiGRgj20XfFj11euX0dmt-DmJ9DlCGltswvbXn6sHf8) [ ](https://private-user-images.githubusercontent.com/3807253/363257115-47e083f9-e6f1-4d6e-a7ce-04cab397a398.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUwMzYxMjUsIm5iZiI6MTcyNTAzNTgyNSwicGF0aCI6Ii8zODA3MjUzLzM2MzI1NzExNS00N2UwODNmOS1lNmYxLTRkNmUtYTdjZS0wNGNhYjM5N2EzOTguZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MzBUMTYzNzA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjEyZDc2YmY0Y2QwYWM3MDdjMDVlYmVhMDg1ZDk2ZTg3NzIxMjIxNWM3YjdiZDhiNjg0M2VmMWFjNDY0Y2M5YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.FiGRgj20XfFj11euX0dmt-DmJ9DlCGltswvbXn6sHf8)
> 
> Also, I should refresh the page before writing a response 😅

That's really cool! With the explicit name no ambiguity. This looks good enough to me.

**Reply by soronpo** [reply]
Since we can solve this from Metals' side, I opened issue #6738
